processor/spanmetricsprocessor/processor.go

```diff
@@ -289,8 +313,26 @@ func (p *processorImp) collectCallMetrics(ilm pdata.InstrumentationLibraryMetric
 		dpCalls.SetTimestamp(pdata.NewTimestampFromTime(time.Now()))
 		dpCalls.SetIntVal(p.callSum[key])
 
-		p.metricKeyToDimensions[key].CopyTo(dpCalls.Attributes())
+		dimensions, err := p.getDimensionsByMetricKey(key)
+		if err != nil {
+			p.logger.Error(err.Error())
```

Comment:
My understanding is that this error will only come about by programmer error. If the logic is correct, this error should _never_ come to being. That is, the items cached should always be of type `pdata.AttributeMap` and the key should _always_ exist in the `metricKeyToDimensions` cache; even if it is evicted, it's still "gettable" but could be purged on the next batch of traces. 

If this is correct, should we strongly assert this by panicking? I may be wrong with my understanding as well 😄

Reply:
Your understanding above is correct. 

I have concerns about handling this error like `panic(err.Error())`.

My points are
- even if this unexpected error happens, we still can't assume this will happen again for processing the next batch of spans.
- this kind of error should not put the whole application/goroutine down.

How do you think?

Reply:
>> even if this unexpected error happens, we still can't assume this will happen again for processing the next batch of spans.
>
>Could you elaborate more on what you mean?

I mean panic should be used for error that is fatal and not recoverable. As we remove the evicted items for cumulative temporality aggregation and purge the whole cache for delta temporality aggregation, there are chances the cache will be reset and the error will not happen again for the next batch of spans.

>IMO I think it should crash the application so it can be fixed sooner. Logging the error can make these bugs easy to miss and lead to incorrect data, which I think is a worse scenario.

I'm not sure if this processor will be executed in the main goroutine. I'm worried about it could be worse to crash the whole application at run time.

>The alternative is to bubble the error up by changing the method signatures of caller methods.
I'll do this change and we can check will that makes sense.

Reply:
Now it's bubbled out as:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6503/commits/393a35862cbdb965682498f2bfcfb4f43737255c

@MovieStoreGuy @jamesmoessis does this make sense to you or would you prefer panic?

Reply:
This doesn't seem to be part of the bootstrap, so, it shouldn't panic:

https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#do-not-crash-after-startup